src/common/runtime/server.ts

```diff
@@ -0,0 +1,117 @@
+/* eslint no-console: "off" */
+
+import * as http from 'http';
+import { AddressInfo } from 'net';
+
+import { DefaultTestFileLoader } from '../internal/file_loader.js';
+import { prettyPrintLog } from '../internal/logging/log_message.js';
+import { Logger } from '../internal/logging/logger.js';
+import { LiveTestCaseResult } from '../internal/logging/result.js';
+import { parseQuery } from '../internal/query/parseQuery.js';
+import { TestQueryWithExpectation } from '../internal/query/query.js';
+import { TestTreeLeaf } from '../internal/tree.js';
+import { setGPUProvider } from '../util/navigator_gpu.js';
+
+import sys from './helper/sys.js';
+
+function usage(rc: number): never {
+  console.log('Usage:');
+  console.log(`  tools/run_${sys.type} [OPTIONS...]`);
+  console.log('Options:');
+  console.log('  --verbose       Print result/log of every test as it runs.');
+  console.log('  --gpu-provider  Path to node module that provides the GPU implementation.');
+  console.log(``);
+  console.log(`Provides an HTTP server used for running tests via an HTTP RPC interface`);
+  console.log(`To run a test, perform an HTTP GET or POST at the URL:`);
+  console.log(`  http://localhost:port/run?<test-name>`);
+  console.log(`To shutdown the server perform an HTTP GET or POST at the URL:`);
+  console.log(`  http://localhost:port/terminate`);
+  return sys.exit(rc);
+}
+
+interface RunResult {
+  status: string;
+  message: string;
+}
+
+if (!sys.existsSync('src/common/runtime/cmdline.ts')) {
+  console.log('Must be run from repository root');
+  usage(1);
+}
+
+let debug = false;
+
+for (let i = 0; i < sys.args.length; ++i) {
+  const a = sys.args[i];
+  if (a.startsWith('-')) {
+    if (a === '--verbose') {
+      debug = true;
+    } else if (a === '--gpu-provider') {
+      const modulePath = sys.args[++i];
+      setGPUProvider(() => require(modulePath).gpu);
+    }
+  }
+}
+
+(async () => {
+  Logger.globalDebugMode = debug;
+  const log = new Logger();
+  const testcases = await allWebGPUTestcases();
```

Comment:
nit: if we just have

`const testcasesPormise = allWebGPUTestcases();`

and later on inside the request handler do:

`const testcase = (await testcasesPromise).get(name);`

then the server will get instantiated and start listening right away, and block later on if testcases aren't done loading yet.
I find it a minor annoyance when I start a tool, hit the web server, and then realize it hasn't started listening yet.